walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,206 @@
+      Collection<Event> events, Collection<String> attendees) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    // If an event's attendee list contains a names of an attendee that needs to attend
+    // (from Collection<String> attendees, then we add the event's TimeRange to a list.
+    events.forEach(
+        event -> {
+          Collection<String> eventAttendees = event.getAttendees();
+
+          for (String attendee : eventAttendees) {
+            if (!attendees.contains(attendee)) {
+              continue;
+            }
+
+            TimeRange currentEventTimeRange = event.getWhen();
+
+            // If the currentEventTimeRange isn't merged into a TimeRange within the
+            // unavailableTimeRanges list, then add the currentEventTimeRange to the list.
+            if (!mergeOnOverlap(unavailableTimeRanges, currentEventTimeRange)) {
+              unavailableTimeRanges.add(currentEventTimeRange);
+            }
+
+            break;
+          }
+        });
+
+    return unavailableTimeRanges;
+  }
+
+  /**
+   * Returns whether a merge has ocurred or not and modifies the Collection of TimeRanges to include
+   * a merged TimeRange where the newTimeRange overlaps with the passed Collection.
+   */
+  private boolean mergeOnOverlap(Collection<TimeRange> timeRanges, TimeRange newTimeRange) {
+    // The default value is set to false, which is converted to true if a merge takes place.
+    boolean doOverlap = false;
+
+    // Store items that we will add and remove from the collection after we finish iterating
+    // through it in order to avoid ConcurrentModificationException from being thrown.
+    List<TimeRange> toRemove = new ArrayList<>();
+    List<TimeRange> toAdd = new ArrayList<>();
+
+    for (TimeRange takenTimeRange : timeRanges) {
+      if (newTimeRange.overlaps(takenTimeRange)) {
+        TimeRange combinedTimeRange = mergeTimeRanges(newTimeRange, takenTimeRange);
+        toRemove.add(takenTimeRange);
+        toAdd.add(combinedTimeRange);
+        doOverlap = true;
+      }
+    }
+
+    for (TimeRange timeRange : toRemove) {
+      timeRanges.remove(timeRange);
+    }
+
+    for (TimeRange timeRange : toAdd) {
+      timeRanges.add(timeRange);
+    }
+
+    return doOverlap;
+  }
+
+  private TimeRange mergeTimeRanges(TimeRange firstTimeRange, TimeRange secondTimeRange) {
+    TimeRange mergedTimeRange =
+        TimeRange.fromStartEnd(
+            Math.min(firstTimeRange.start(), secondTimeRange.start()),
+            Math.max(firstTimeRange.end(), secondTimeRange.end()),
+            /* inclusive = */ false);
+
+    return mergedTimeRange;
+  }
+
+  private List<TimeRange> getAvailableTimeRanges(
+      List<TimeRange> unavailableTimeRanges, long durationMinutes) {
+    if (unavailableTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collections.sort(unavailableTimeRanges, TimeRange.ORDER_BY_START);
```

Comment:
It's good practice to make a defensive copy of an argument object before you mutate it unless that's explicitly the intent of the method (and you document it in a comment). In this case, the caller passed in `unavailableTimeRanges` but may not know that the method will sort it. If they try to use it for something else later it could have unexpected consequences.

Reply:
Done.